### PR TITLE
filters: allow envoy filters to be created as exception free

### PR DESCRIPTION
**Pull request by alyssawilk**
I tried my best here to keep churn minimal, so the default APIs are the same for almost all cc code (though not for tests) and you opt into exceptionless filters.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

https://github.com/envoyproxy/envoy-mobile/issues/176

### CODE REVIEW

**Review by RyanTheOptimist — DISMISSED** [review]
Wow, awesome! Would you mind updating the PR description briefly to explain the new class name and factory API that this PR introduces?

File: test/extensions/filters/http/admission_control/config_test.cc [hunk 1]
```diff
@@ -75,8 +75,10 @@ sampling_window: 1337s
   TestUtility::loadFromYamlAndValidate(yaml, proto);
   NiceMock<Server::Configuration::MockFactoryContext> factory_context;
   EXPECT_THROW_WITH_MESSAGE(
-      admission_control_filter_factory.createFilterFactoryFromProtoTyped(
-          proto, "whatever", dual_info_, factory_context.getServerFactoryContext()),
+      EXPECT_TRUE(admission_control_filter_factory
+                      .createFilterFactoryFromProtoTyped(proto, "whatever", dual_info_,
+                                                         factory_context.getServerFactoryContext())
+                      .ok()),
```
Comment (ravenblackx) [inline test/extensions/filters/http/admission_control/config_test.cc:81]: I don't understand how this construct is working? `EXPECT_TRUE` doesn't throw an exception does it? And if the inner part threw an exception then `.ok()` wouldn't be true so it seems like it ought to fail either way?
Even if it does work, I think the concept would be more readable like
```
EXPECT_THAT(admission_control_filter_factory
                      .createFilterFactoryFromProtoTyped(proto, "whatever", dual_info_,
                                                         factory_context.getServerFactoryContext()), 
    HasStatusMessage(HasSubstr("Success rate threshold cannot be less than 1.0%.")));
```
Or, if the thing is that it's throwing an exception and the EXPECT_TRUE is being ignored, then
```
  EXPECT_THROW_WITH_MESSAGE(
      admission_control_filter_factory
                      .createFilterFactoryFromProtoTyped(proto, "whatever", dual_info_,
                                                         factory_context.getServerFactoryContext())
                      .IgnoreResult(),
    EnvoyException, "Success rate threshold cannot be less than 1.0%.");
```

Comment (alyssawilk) [reply test/extensions/filters/http/admission_control/config_test.cc:81]: the assert is being ignored. 
 no member named 'IgnoreResult' in 'absl::StatusOr<std::function<void (Envoy::Http::FilterChainFactoryCallbacks &)>>' ditto absl::Status  not sure what you're suggesting I use?

File: test/extensions/filters/http/aws_lambda/config_test.cc [hunk 2]
```diff
@@ -110,7 +112,7 @@ arn: "arn:aws:lambda:region:424242:fun"
   testing::NiceMock<Server::Configuration::MockFactoryContext> context;
   AwsLambdaFilterFactory factory;
 
-  EXPECT_THROW(factory.createFilterFactoryFromProto(proto_config, "stats", context),
+  EXPECT_THROW(factory.createFilterFactoryFromProto(proto_config, "stats", context).value(),
```
Comment (wbpcode) [inline test/extensions/filters/http/aws_lambda/config_test.cc:115]: nit: `status().IgnoreError()`?

File: test/extensions/filters/http/admission_control/config_test.cc [hunk 3]
```diff
@@ -103,8 +106,10 @@ sampling_window: 1337s
   TestUtility::loadFromYamlAndValidate(yaml, proto);
   NiceMock<Server::Configuration::MockFactoryContext> factory_context;
   EXPECT_THROW_WITH_MESSAGE(
-      admission_control_filter_factory.createFilterFactoryFromProtoTyped(
-          proto, "whatever", dual_info_, factory_context.getServerFactoryContext()),
+      admission_control_filter_factory
+          .createFilterFactoryFromProtoTyped(proto, "whatever", dual_info_,
+                                             factory_context.getServerFactoryContext())
+          .IgnoreError(),
```
Comment (wbpcode) [inline test/extensions/filters/http/admission_control/config_test.cc:112]: nit: `status().IgnoreError()`?

File: test/extensions/filters/http/cache/config_test.cc [hunk 4]
```diff
@@ -44,13 +46,16 @@ TEST_F(CacheFilterFactoryTest, Disabled) {
...
+      factory_.createFilterFactoryFromProto(config_, "stats", context_).status().IgnoreError(),
+      EnvoyException);
 }
 
 TEST_F(CacheFilterFactoryTest, UnregisteredTypedConfig) {
   config_.mutable_typed_config()->PackFrom(
       envoy::extensions::filters::http::cache::v3::CacheConfig());
-  EXPECT_THROW(factory_.createFilterFactoryFromProto(config_, "stats", context_), EnvoyException);
+  EXPECT_THROW(factory_.createFilterFactoryFromProto(config_, "stats", context_).IgnoreError(),
+               EnvoyException);
```
Comment (wbpcode) [inline test/extensions/filters/http/cache/config_test.cc:58]: ditto.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

Seems there are three type of ways to handle the return value when we expect a exception. Call the`value()` directly, call the `IgnoreError()` directly, call the `status().IgnoreError()`. Is is a ignorance or by design? (But it's OK for me if CI is passed. I think we will gradually revise these exception to error message checking?)

**Pull request merged**